### PR TITLE
CSHARP-5439: Fix flaky CSFLE tests

### PR DESCRIPTION
**Pull request by sanych-sun**

### CODE REVIEW

File: evergreen/evergreen.yml [hunk 1]
```diff
@@ -2684,7 +2684,7 @@ buildvariants:
     - name: test-csfle-with-mongocryptd-net60
 
 - matrix_name: "csfle-with-azure-kms-tests-linux"
-  matrix_spec: { ssl: "nossl", os: "ubuntu-1804" }
+  matrix_spec: { ssl: "nossl", os: "ubuntu-2004" }
```
Comment (sanych-sun) [inline evergreen/evergreen.yml:2687]: This is not really related to the fix, just decided to bump the OS as well.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 2]
```diff
@@ -2524,36 +2507,28 @@ private void AddOrReplace<TValue>(IDictionary<string, TValue> dict, string key,
...
-                e.Message.Should().Contain(innerMessage);
-                if (e.InnerException != null)
-                {
-                    e = e.InnerException;
-                }
+                e = e.InnerException;
             }
 
-            if (typeof(OperationCanceledException).IsAssignableFrom(exType))
+            if (e == null)
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:2519]: remove this debug code

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 3]
```diff
@@ -1825,46 +1820,27 @@ void AssertException(Exception ex)
...
-                                        break;
-                                    default: throw new Exception($"Unexpected OS: {currentOperatingSystem}");
-                                }
+                                AssertInnerEncryptionException<MongoClientException>(ex, e => e.Message.Should().Contain("Failed to acquire IMDS access token."));
                             }
                             break;
                         case "gcp":
                             {
-                                AssertInnerEncryptionException<HttpRequestException>(ex, "Failed to acquire gce metadata credentials.");
+                                AssertInnerEncryptionException<MongoClientException>(ex, e => e.Message.Should().Contain("Failed to acquire gce metadata credentials."));
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:1843]: Consider an overload that just accepts the message?

**Review by BorisDog — APPROVED** [review]
LGTM + small suggestion + don't forget to remove the debug code.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 4]
```diff
@@ -1657,50 +1668,29 @@ void AssertTlsWithoutClientCertOnLinux(Exception exception)
...
                 }
             }
 
             void AssertTlsWithoutClientCertOnWindows(Exception exception)
             {
                 try
                 {
-                    string[] innerExceptions =
-#if NET6_0_OR_GREATER
-                        ["Authentication failed because the remote party sent a TLS alert"];
```
Comment (adelinowona) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:1688]: Is this error message impossible to get anymore when using Net6 on windows?

**Review by adelinowona — APPROVED** [review]
LGTM! Nice refactoring on the error message assertion mechanism! I just left a single comment.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 4 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:1688]: The real exception thrown in this situation looks like:
```
System.Net.Sockets.SocketException: Encryption related exception: Authentication failed because the remote party sent a TLS alert: 'HandshakeFailure'.
 ---> System.Security.Authentication.AuthenticationException: Authentication failed because the remote party sent a TLS alert: 'HandshakeFailure'.
 ---> System.ComponentModel.Win32Exception (0x80090326): The message received was unexpected or badly formatted.
```

So now as we decided not to analyze all exceptions, but looking for very specific one in the chain of inner exceptions, this turns to be a general case to look for Win32Exception with message " The message received was unexpected or badly formatted.".
